source/dofs/dof_handler_policy.cc

```diff
@@ -3957,128 +3974,73 @@ namespace internal
+        // First figure out the new set of locally owned DoF indices.
+        // If we own no DoFs, we still need to go through this function,
+        // but we can skip this calculation.
+        //
+        // The IndexSet::add_indices() function is substantially more
+        // efficient if the set of indices is already sorted because
+        // it can then insert ranges instead of individual elements.
+        // consequently, pre-sort the array of new indices
+        IndexSet my_locally_owned_dofs (dof_handler->n_dofs());
+        if (dof_handler->n_locally_owned_dofs() > 0)
           {
-            std::vector<dealii::types::global_dof_index> new_numbers_sorted (new_numbers);
+            std::vector<dealii::types::global_dof_index> new_numbers_sorted = new_numbers;
             std::sort(new_numbers_sorted.begin(), new_numbers_sorted.end());
-            std::vector<dealii::types::global_dof_index>::const_iterator it = new_numbers_sorted.begin();
-            const unsigned int n_blocks = dof_handler->get_fe().n_blocks();
-            std::vector<std::pair<dealii::types::global_dof_index,unsigned int> > block_indices(n_blocks);
-            block_indices[0].first = *it++;
-            block_indices[0].second = 1;
-            unsigned int current_block = 0, n_filled_blocks = 1;
-            for ( ; it != new_numbers_sorted.end(); ++it)
-              {
-                bool done = false;
 
-                // search from the current block onwards whether the next
-                // index is shifted by one from the previous one.
-                for (unsigned int i=0; i<n_filled_blocks; ++i)
-                  if (*it == block_indices[current_block].first
-                      +block_indices[current_block].second)
-                    {
-                      block_indices[current_block].second++;
-                      done = true;
-                      break;
-                    }
-                  else
-                    {
-                      if (current_block == n_filled_blocks-1)
-                        current_block = 0;
-                      else
-                        ++current_block;
-                    }
-
-                // could not find any contiguous range: need to add a new
-                // block if possible. Abort otherwise, which will add all
-                // elements individually to the IndexSet.
-                if (done == false)
-                  {
-                    if (n_filled_blocks < n_blocks)
-                      {
-                        block_indices[n_filled_blocks].first = *it;
-                        block_indices[n_filled_blocks].second = 1;
-                        current_block = n_filled_blocks;
-                        ++n_filled_blocks;
-                      }
-                    else
-                      break;
-                  }
-              }
+            my_locally_owned_dofs.add_indices (new_numbers_sorted.begin(),
```

Comment:
The reason for skipping all these calculations is that `IndexSet::add_indices` merges ranges on its own, isn't it?